src/Security/LdapUserProvider.php

```diff
@@ -0,0 +1,199 @@
+        // If user has attribute, assign roles that map
+        $isMemberOfValues = $entry->getAttribute('isMemberOf');
+        $hasRoles = array_intersect_key($this->roleMapping, array_flip($isMemberOfValues));
+        $roles = array_merge($this->defaultRoles, $hasRoles);
+
+        if (count($roles) === 0) {
+            throw new UsernameNotFoundException('You do not have permission to use this application');
+        }
+
+        if (!$this->supportsClass(new $this->userClass())) {
+            throw new \RuntimeException('Userclass must be of type ' . LdapUser::class . ' or a child class.');
+        }
+
+        /** @var LdapUser $user */
+        $user = new $this->userClass(
+            $this->getAttributeValue($entry, 'uid'),
+            $displayName ?? $uid,
+            $roles
+        );
+
+        return $user;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadUserByUsername($username)
+    {
+        try {
+            $this->ldap->bind($this->searchDn, $this->searchPassword);
+            $username = $this->ldap->escape($username, '', LdapInterface::ESCAPE_FILTER);
+            $query = str_replace('{username}', $username, $this->defaultSearch);
+            $search = $this->ldap->query($this->baseDn, $query);
+        } catch (ConnectionException $e) {
+            throw new UsernameNotFoundException(sprintf('User "%s" not found.', $username), 0, $e);
+        }
+
+        $entries = $search->execute();
+        $count = \count($entries);
+
+        if (!$count) {
+            throw new UsernameNotFoundException(sprintf('User "%s" not found.', $username));
+        }
+
+        if ($count > 1) {
+            throw new UsernameNotFoundException('More than one user found');
+        }
+
+        $entry = $entries[0];
+
+        $username = $this->getAttributeValue($entry, 'uid');
+
+        return $this->loadUser($username, $entry);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function refreshUser(UserInterface $user)
+    {
+        if (!$this->supportsClass($user)) {
+            throw new UnsupportedUserException(sprintf('Instances of "%s" are not supported.', \get_class($user)));
+        }
+
+        return new $this->userClass($user->getUsername(), $user->getDisplayName(), $user->getRoles());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsClass($class)
+    {
+        return $class instanceof LdapUser;
+    }
+
+    /**
+     * Fetches a required unique attribute value from an LDAP entry.
+     *
+     * @param Entry|null $entry
+     * @param string $attribute
+     * @return mixed
+     */
+    private function getAttributeValue(Entry $entry, $attribute)
+    {
+        if (!$entry->hasAttribute($attribute)) {
+            throw new InvalidArgumentException(sprintf('Missing attribute "%s" for user "%s".', $attribute, $entry->getDn()));
+        }
+
+        $values = $entry->getAttribute($attribute);
+
+        if (1 !== \count($values)) {
```

Comment:
I am not against yoda style but I would either use regular or yoda style all the time.